File: android/src/main/java/com/smileidentity/react/expo/SmileIDBiometricKYCView.kt
package com.smileidentity.react.expo

import android.content.Context
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.systemBarsPadding
import androidx.compose.runtime.Composable
import androidx.compose.runtime.mutableStateOf
import androidx.compose.ui.Modifier
import com.smileidentity.SmileID
import com.smileidentity.compose.BiometricKYC
import com.smileidentity.models.ConsentInformation
import com.smileidentity.models.IdInfo
import com.smileidentity.models.SmileSensitivity
import com.smileidentity.results.BiometricKycResult
import com.smileidentity.results.SmileIDResult
import com.smileidentity.util.randomJobId
import com.smileidentity.util.randomUserId
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher
import kotlinx.collections.immutable.ImmutableMap
import kotlinx.collections.immutable.persistentMapOf

/**
 * Biometric KYC View using ExpoView
 **/
class SmileIDBiometricKYCView(context: Context, appContext: AppContext) : SmileIDExpoComposeView(
    context = context,
    appContext = appContext,
    shouldHostComposeContent = true
) {
    private var props = mutableStateOf(BiometricKYCProps())
    private val onResult by EventDispatcher()
    private val onError by EventDispatcher()

    @Composable
    override fun Content() {
        BiometricKYCView(
            props = props.value,
            onResult = { result ->
                onResult(
                    mapOf(
                        "selfieFile" to result.selfieFile.toString(),
                        "livenessFiles" to result.livenessFiles.toString(),
                        "didSubmitBiometricJob" to result.didSubmitBiometricKycJob
                    )
                )
            },
            onError = { error ->
                onError(
                    mapOf(
                        "error" to error.localizedMessage.toString()
                    )
                )
            }
        )
    }

    fun updateConfig(config: BiometricKYCParams) {
        props.value = config.toBiometricKYCProps()
    }
}

/**
 * Compose view that wraps the SmileID Biometric KYV screen
 **/
@Composable
private fun BiometricKYCView(
    props: BiometricKYCProps,
    onResult: (BiometricKycResult) -> Unit,
    onError: (Throwable) -> Unit
) {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .navigationBarsPadding()
            .systemBarsPadding()
    ) {
        SmileID.BiometricKYC(
            userId = props.userId ?: randomUserId(),
            jobId = props.jobId ?: randomJobId(),
            allowNewEnroll = props.allowNewEnroll,
            allowAgentMode = props.allowAgentMode,
            showAttribution = props.showAttribution,
            showInstructions = props.showInstructions,
            smileSensitivity = props.smileSensitivity,
            extraPartnerParams = props.extraParams,
            consentInformation = props.consentInformation,
            idInfo = props.idInfo
        ) { result ->
            when(result) {
                is SmileIDResult.Success -> {
                    onResult(result.data)
                }
                is SmileIDResult.Error -> {
                    onError(result.throwable)
                }
            }
        }
    }
}

data class BiometricKYCProps(
    val userId: String? = null,
    val jobId: String? = null,
    val allowNewEnroll: Boolean = true,
    val allowAgentMode: Boolean = false,
    val showAttribution: Boolean = true,
    val showInstructions: Boolean = true,
    val skipApiSubmission: Boolean = false,
    val smileSensitivity: SmileSensitivity = SmileSensitivity.NORMAL,
    val useStrictMode: Boolean = false,
    val extraParams: ImmutableMap<String, String> = persistentMapOf(),
    val consentInformation: ConsentInformation ? = null,
    val idInfo: IdInfo = IdInfo(
        country = ""
    )
)

File: ios/SmileIDBiometricKYCView.swift
import ExpoModulesCore
import SmileID
import SwiftUI

// Biometric KYC View using ExpoView
final class SmileIDBiometricKYCView: ExpoView {
    let onResult = EventDispatcher()
    let onError = EventDispatcher()
    private let delegate: BiometricKYCDelegate
    private let hostingController: UIHostingController<BiometricKYCView>
    private var config: BiometricKYCParams?

    required init(appContext: AppContext? = nil) {
        delegate = BiometricKYCDelegate()
        hostingController = UIHostingController(
            rootView: BiometricKYCView(
                delegate: delegate,
                config: nil
            )
        )
        // Force light mode
        hostingController.overrideUserInterfaceStyle = .light

        super.init(appContext: appContext)

        // Set up delegate callbacks
        delegate.onResult = { [weak self] result in
            self?.onResult(result)
        }

        delegate.onError = { [weak self] error in
            self?.onError(["error": error.localizedDescription])
        }

        // Add the hosting controller's view
        addSubview(hostingController.view)
        hostingController.view.fillSuperview()
    }

    func updateConfig(_ config: BiometricKYCParams) {
        self.config = config
        hostingController.rootView = BiometricKYCView(
            delegate: delegate,
            config: config
        )
    }
}

// SwiftUI view that wraps the SmileID biometric kyc screen
struct BiometricKYCView: View {
    let delegate: BiometricKYCDelegate
    let config: BiometricKYCParams?
    
    var body: some View {
        if let config = config {
            SmileID.biometricKycScreen(
                idInfo: config.idInfo.toIdInfo(),
                userId: config.userId ?? generateUserId(),
                jobId: config.jobId ?? generateJobId(),
                allowNewEnroll: config.allowNewEnroll,
                allowAgentMode: config.allowAgentMode,
                forceAgentMode: config.forceAgentMode,
                showAttribution: config.showAttribution,
                showInstructions: config.showInstructions,
                smileSensitivity: SmileSensitivity(from: config.smileSensitivity),
                useStrictMode: config.useStrictMode,
                skipApiSubmission: config.skipApiSubmission,
                extraPartnerParams: config.extraPartnerParams,
                consentInformation: config.consentInformation?.toConsentInformation(),
                delegate: delegate
            )
        } else {
            Text("Configuration not provided").foregroundColor(.red)
        }
    }
}

// Delegate class for biometric KYC
class BiometricKYCDelegate: BiometricKycResultDelegate {
    var onResult: (([String: Any]) -> Void)?
    var onError: ((Error) -> Void)?

    func didSucceed(
        selfieImage: URL,
        livenessImages: [URL],
        didSubmitBiometricJob: Bool
    ) {
        let result: [String: Any] = [
            "selfieImage": selfieImage.absoluteString,
            "livenessImages": livenessImages.map { $0.absoluteString },
            "didSubmitBiometricJob": didSubmitBiometricJob
        ]
        onResult?(result)
    }

    func didError(error: Error) {
        onError?(error)
    }
}
